File: pkg/models/category_budget_limit.go
package models

// [PLUGIN:budget] This file implements the category budget limit feature as a self-contained
// extension. It does NOT modify the core Transaction model. See docs/PLUGIN_DESIGN.md.

// CategoryBudgetLimit represents a monthly spending limit set on a transaction category.
// One row per (uid, category_id, start_date). Mirrors Firefly III's budget_limits concept,
// simplified: no rollover, no join table (the link to transactions is implicit through
// Transaction.CategoryId, which already exists in the core model).
type CategoryBudgetLimit struct {
	BudgetId        int64  `xorm:"PK"`
	Uid             int64  `xorm:"INDEX(IDX_category_budget_uid_deleted) NOT NULL unique(UQE_category_budget_uid_category_startdate)"`
	CategoryId      int64  `xorm:"NOT NULL unique(UQE_category_budget_uid_category_startdate)"`
	StartDate       int64  `xorm:"NOT NULL unique(UQE_category_budget_uid_category_startdate)"` // unix time at 00:00:00 of the 1st of the month (client timezone)
	EndDate         int64  `xorm:"NOT NULL"`                                                  // unix time at 23:59:59 of the last day of the month (client timezone)
	Amount          int64  `xorm:"NOT NULL"`                                                  // limit in minor units (cents)
	Currency        string `xorm:"VARCHAR(8) NOT NULL"`                                       // base currency code for v1
	Deleted         bool   `xorm:"INDEX(IDX_category_budget_uid_deleted) NOT NULL"`
	CreatedUnixTime int64
	UpdatedUnixTime int64
	DeletedUnixTime int64
}

// CategoryBudgetLimitListByMonthRequest represents all parameters of category budget limit listing request
type CategoryBudgetLimitListByMonthRequest struct {
	// StartDate is the unix time at 00:00:00 of the 1st of the target month (client timezone)
	StartDate int64 `form:"startDate,string" binding:"required,min=1"`
}

// CategoryBudgetLimitGetRequest represents all parameters of a single category budget limit getting request
type CategoryBudgetLimitGetRequest struct {
	Id int64 `form:"id,string" binding:"required,min=1"`
}

// CategoryBudgetLimitCreateRequest represents all parameters of a single category budget limit creation request
type CategoryBudgetLimitCreateRequest struct {
	CategoryId      int64  `json:"categoryId,string" binding:"required,min=1"`
	StartDate       int64  `json:"startDate,string" binding:"required,min=1"` // unix time at 00:00:00 of the 1st of the month
	Amount          int64  `json:"amount" binding:"required,validTransactionAmount"`
	Currency        string `json:"currency" binding:"required,max=8"`
	ClientSessionId string `json:"clientSessionId"`
}

// CategoryBudgetLimitModifyRequest represents all parameters of a single category budget limit modification request
type CategoryBudgetLimitModifyRequest struct {
	Id         int64 `json:"id,string" binding:"required,min=1"`
	Amount     int64 `json:"amount" binding:"required,validTransactionAmount"`
	Currency   string `json:"currency" binding:"required,max=8"`
}

// CategoryBudgetLimitDeleteRequest represents all parameters of a single category budget limit deletion request
type CategoryBudgetLimitDeleteRequest struct {
	Id int64 `json:"id,string" binding:"required,min=1"`
}

// CategoryBudgetLimitInfoResponse represents a view-object of a single category budget limit
type CategoryBudgetLimitInfoResponse struct {
	Id              int64  `json:"id,string"`
	CategoryId      int64  `json:"categoryId,string"`
	StartDate       int64  `json:"startDate,string"`
	EndDate         int64  `json:"endDate,string"`
	Amount          int64  `json:"amount"`
	Currency        string `json:"currency"`
}

// ToCategoryBudgetLimitInfoResponse returns a view-object according to the database model
func (b *CategoryBudgetLimit) ToCategoryBudgetLimitInfoResponse() *CategoryBudgetLimitInfoResponse {
	return &CategoryBudgetLimitInfoResponse{
		Id:         b.BudgetId,
		CategoryId: b.CategoryId,
		StartDate:  b.StartDate,
		EndDate:    b.EndDate,
		Amount:     b.Amount,
		Currency:   b.Currency,
	}
}

// CategoryBudgetOverviewItem represents one row of the budget overview for a given month:
// the limit set on a category plus how much has actually been spent on that category in the period.
type CategoryBudgetOverviewItem struct {
	*CategoryBudgetLimitInfoResponse
	CategoryName      string `json:"categoryName"`
	CategoryParentId  int64  `json:"categoryParentId,string"`
	CategoryType      TransactionCategoryType `json:"categoryType"`
	ActualExpenseAmount int64 `json:"actualExpenseAmount"` // total spent in the period, minor units (already base currency)
	AvailableAmount   int64 `json:"availableAmount"`       // Amount - ActualExpenseAmount (can be negative)
}

// CategoryBudgetOverviewResponse represents the budget overview for a given month
type CategoryBudgetOverviewResponse struct {
	StartDate       int64                        `json:"startDate,string"`
	EndDate         int64                        `json:"endDate,string"`
	TotalLimit      int64                        `json:"totalLimit"`
	TotalActual     int64                        `json:"totalActualExpenseAmount"`
	TotalAvailable  int64                        `json:"totalAvailableAmount"`
	Items           []*CategoryBudgetOverviewItem `json:"items"`
}

File: src/models/category_budget_limit.ts
// [PLUGIN:budget] Category budget limit feature - TypeScript model.
// Mirrors the backend DTOs in pkg/models/category_budget_limit.go.

export class CategoryBudgetLimit implements CategoryBudgetLimitInfoResponse {
    public id: string;
    public categoryId: string;
    public startDate: number;
    public endDate: number;
    public amount: number;
    public currency: string;

    public constructor(id: string, categoryId: string, startDate: number, endDate: number, amount: number, currency: string) {
        this.id = id;
        this.categoryId = categoryId;
        this.startDate = startDate;
        this.endDate = endDate;
        this.amount = amount;
        this.currency = currency;
    }

    public toCreateRequest(clientSessionId: string): CategoryBudgetLimitCreateRequest {
        return {
            categoryId: this.categoryId,
            // startDate is sent as a quoted string: the backend CreateRequest uses the
            // `,string` json tag (consistent with all large int64 fields in ezBookkeeping).
            startDate: String(this.startDate),
            amount: this.amount,
            currency: this.currency,
            clientSessionId
        };
    }

    public toModifyRequest(): CategoryBudgetLimitModifyRequest {
        return {
            id: this.id,
            amount: this.amount,
            currency: this.currency
        };
    }

    public static of(resp: CategoryBudgetLimitInfoResponse): CategoryBudgetLimit {
        return new CategoryBudgetLimit(resp.id, resp.categoryId, resp.startDate, resp.endDate, resp.amount, resp.currency);
    }

    public static ofMulti(responses: CategoryBudgetLimitInfoResponse[]): CategoryBudgetLimit[] {
        return responses.map(CategoryBudgetLimit.of);
    }

    public static createNew(categoryId: string, startDate: number, currency: string): CategoryBudgetLimit {
        return new CategoryBudgetLimit('', categoryId, startDate, 0, 0, currency);
    }
}

export interface CategoryBudgetLimitCreateRequest {
    readonly categoryId: string;
    readonly startDate: string;
    readonly amount: number;
    readonly currency: string;
    readonly clientSessionId: string;
}

export interface CategoryBudgetLimitModifyRequest {
    readonly id: string;
    readonly amount: number;
    readonly currency: string;
}

export interface CategoryBudgetLimitListByMonthRequest {
    readonly startDate: string;
}

export interface CategoryBudgetLimitInfoResponse {
    readonly id: string;
    readonly categoryId: string;
    readonly startDate: number;
    readonly endDate: number;
    readonly amount: number;
    readonly currency: string;
}

export interface CategoryBudgetOverviewItem extends CategoryBudgetLimitInfoResponse {
    readonly categoryName: string;
    readonly categoryParentId: string;
    readonly categoryType: number;
    readonly actualExpenseAmount: number;
    readonly availableAmount: number;
}

export interface CategoryBudgetOverviewResponse {
    readonly startDate: number;
    readonly endDate: number;
    readonly totalLimit: number;
    readonly totalActualExpenseAmount: number;
    readonly totalAvailableAmount: number;
    readonly items: CategoryBudgetOverviewItem[];
}
